feat(sync): make SyncerRegistry thread-safe and execute syncers concurrently

powerslider opened the pull request:
## Why this should be merged

This change makes the registry production-ready for concurrent environments where multiple goroutines may register syncers simultaneously, and ensures all syncers execute concurrently for better performance.

## How this works

- Replace slice-based storage with sync.Map for concurrent access.
- Use LoadOrStore for atomic duplicate prevention during registration.
- Execute all registered syncers concurrently using errgroup.
- Update tests to work with concurrent model and verify thread safety.

### Implement Cancellation-Aware Result 
- Add `synccommon.WaitResult` type to unify return values for both successful completion, clean cancellation, and real errors.
- Implement `WaitForCompletion(ctx, doneCh, cancelFunc)` helper to DRY up cancellation versus completion logic across syncers.
- Refactor `stateSyncer.Wait` and `atomicSyncer.Wait` to use the new helper and return `WaitResult` instead of raw `error`.
- Update `RunSyncerTasks` to interpret `WaitResult` semantics:
  - Treat `Cancelled=true` or `Err=context.Canceled` as clean aborts.
  - Fail fast on non-cancellation errors.
- Adjust tests to validate the three-way (`Success` / `Cancelled` / `Error`) contract exposed by `WaitResult`.

## How this was tested

Lots of tests were added to `sync/vm/registry_test.go` to cover all possible concurrency edge cases.

## Need to be documented?
no

## Need to update RELEASES.md?
no

resolves #1082

Signed-off-by: Tsvetan Dimitrov (tsvetan.dimitrov@avalabs.org)

powerslider committed:
feat(sync): make SyncerRegistry thread-safe and execute syncers concurrently

- Replace slice-based storage with sync.Map for concurrent access
- Use LoadOrStore for atomic duplicate prevention during registration
- Execute all registered syncers concurrently using errgroup
- Remove manual mutex management in favor of built-in thread safety
- Update tests to work with concurrent model and verify thread safety
- Add TestSyncerRegistry_ConcurrentRegistration to validate concurrent registration
- Replace unrealistic registry reuse test with valuable thread-safety test

This change makes the registry production-ready for concurrent environments
where multiple goroutines may register syncers simultaneously, and ensures
all syncers execute concurrently for better performance.

resolves #1082

Signed-off-by: Tsvetan Dimitrov (tsvetan.dimitrov@avalabs.org)

alarso16 commented on sync/vm/registry_test.go:
```diff
@@ -113,10 +132,31 @@ func TestSyncerRegistry_Register(t *testing.T) {
+			}{
+				{"Syncer1", nil},
+			},
+			expectedCount: 1, // Nil syncer should be allowed
```
Should it be allowed?

alarso16 commented on sync/vm/registry_test.go:
```diff
@@ -260,7 +422,39 @@ func TestSyncerRegistry_RunSyncerTasks(t *testing.T) {
 }
+
+// TestSyncerRegistry_ConcurrentRegistration tests that the registry handles concurrent registration safely.
+func TestSyncerRegistry_ConcurrentRegistration(t *testing.T) {
```
I don't think `SyncerRegistry` needs to be thread-safe. All the initialization should happen on a single thread anyway, so no need for `sync.Map`

alarso16 reviewed (commented):
Didn't fully review the tests, but I'm not sure I agree with the strategy here. `SyncerRegistry` doesn't need to be thread-safe, and if one of the syncers fail, we should immediately exit

powerslider committed:
fix: improve client summary validation and logging

powerslider committed:
fix: disallow passing a nil syncer to be registered

powerslider committed:
fix: disallow running Register after RunSyncerTasks has been started and also running RunSyncerTask a second time

powerslider commented on sync/vm/registry.go:
```diff
@@ -17,56 +19,84 @@ type SyncerTask struct {
 // Returns an error if a syncer with the same name is already registered.
 func (r *SyncerRegistry) Register(name string, syncer synccommon.Syncer) error {
-	if r.registeredNames[name] {
+	task := SyncerTask{name, syncer}
```
That's a great catch. I already fixed this.

powerslider committed:
Merge branch 'master' into powerslider/1082-concurrent-syncer-start

powerslider committed:
feat(sync): implement two-phase start and wait execution

- Introduce a clear “Start Phase” and “Wait Phase” in RunSyncerTasks:
    - Phase 1: concurrently invoke Start() on all registered syncers and abort immediately on any failure.
    - Phase 2: once all Start() calls succeed, concurrently invoke Wait() on each syncer to completion.
- Guarantees atomic bootstrap, fail-fast semantics, and true parallel execution.

Signed-off-by: Tsvetan Dimitrov (tsvetan.dimitrov@avalabs.org)

powerslider committed:
feat(sync): introduce WaitResult abstraction for syncers

- Add `synccommon.WaitResult` type to unify return values for both
  successful completion, clean cancellation, and real errors.
- Implement `WaitForCompletion(ctx, doneCh, cancelFunc)` helper to DRY
  up cancellation versus completion logic across syncers.
- Refactor `stateSyncer.Wait` and `atomicSyncer.Wait` to use the new
  helper and return `WaitResult` instead of raw `error`.
- Update `RunSyncerTasks` to interpret `WaitResult` semantics:
  - Treat `Cancelled=true` or `Err=context.Canceled` as clean aborts.
  - Fail fast on non-cancellation errors.
- Adjust tests to validate the three-way (`Success` / `Cancelled` / `Error`)
  contract exposed by `WaitResult`.

resolves #1082

Signed-off-by: Tsvetan Dimitrov (tsvetan.dimitrov@avalabs.org)

powerslider committed:
Merge branch 'master' into powerslider/1082-concurrent-syncer-start

powerslider committed:
fix: some faulty tests

powerslider commented on sync/vm/registry.go:
```diff
@@ -17,57 +29,106 @@ type SyncerTask struct {
+	// Use reflection to check for nil because in Go, a nil concrete type is not equal to a nil interface [synccommon.Syncer].
+	// When a nil concrete type is assigned to an interface, the interface contains type information even though the value is nil.
+	// reflect.ValueOf(syncer).IsNil() properly detects nil concrete types.
+	if syncer == nil || reflect.ValueOf(syncer).IsNil() {
```
You can check here for explanations: 
https://mangatmodi.medium.com/go-check-nil-interface-the-right-way-d142776edef1

alarso16 commented on sync/types.go:
```diff
@@ -50,3 +63,32 @@ type Extender interface {
+		// Caller cancelled: tell the syncer to shut down.
+		cancelFunc()
+		// Wait for the syncer to finish.
+		<-doneCh
```
This isn't correct - there's a race between the context cancelling and the the syncer finished
